source/AmmoDisplay.cpp

```diff
@@ -0,0 +1,148 @@
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "AmmoDisplay.h"
+
+#include "Color.h"
+#include "text/Font.h"
+#include "text/FontSet.h"
+#include "GameData.h"
+#include "Outfit.h"
+#include "PlayerInfo.h"
+#include "Point.h"
+#include "Rectangle.h"
+#include "Ship.h"
+#include "Sprite.h"
+#include "SpriteSet.h"
+#include "SpriteShader.h"
+
+using namespace std;
+
+
+
+AmmoDisplay::AmmoDisplay(PlayerInfo &player)
+	: player(player)
+{
+}
+
+
+
+void AmmoDisplay::Update(const Ship &flagship)
+{
+	ammo.clear();
+	for(const auto &it : flagship.Weapons())
+	{
+		const Outfit *secWeapon = it.GetOutfit();
+		if(!secWeapon || !secWeapon->Icon() || ammo.find(secWeapon) != ammo.end())
+			continue;
+
+		double ammoCount = -1;
+		// TODO: if a weapon has both an ammo requirement and consumes fuel, the ammo display will only show the ammo,
+		// not the fuel, so the weapon may not fire because it is out of fuel even though it still has ammo,
+		// and will not show a '0' in the ammo display.
+		if(secWeapon->Ammo())
+			ammoCount = flagship.OutfitCount(secWeapon->Ammo());
+		else if(secWeapon->FiringFuel())
+		{
+			double remaining = flagship.Fuel()
+				* flagship.Attributes().Get("fuel capacity");
+			ammoCount = remaining / secWeapon->FiringFuel();
+		}
+		ammo[secWeapon] = ammoCount;
+	}
+}
+
+
+
+void AmmoDisplay::Draw(const Rectangle &ammoBox, const Point &iconDim) const
+{
+	const Font &font = FontSet::Get(14);
+	ammoIconZones.clear();
+
+	const double &ammoIconWidth = iconDim.X();
+	const double &ammoIconHeight = iconDim.Y();
+	// Pad the ammo list by the same amount on all four sides.
+	double ammoPad = .5 * (ammoBox.Width() - ammoIconWidth);
+	const Sprite *selectedSprite = SpriteSet::Get("ui/ammo selected");
+	const Sprite *unselectedSprite = SpriteSet::Get("ui/ammo unselected");
+	const Color &selectedColor = *GameData::Colors().Get("bright");
+	const Color &unselectedColor = *GameData::Colors().Get("dim");
+
+	// This is the bottom left corner of the ammo display.
+	auto pos = Point(ammoBox.Left() + ammoPad, ammoBox.Bottom() - ammoPad);
+	// These offsets are relative to that corner.
+	auto boxOff = Point(ammoIconWidth - .5 * selectedSprite->Width(), .5 * ammoIconHeight);
+	auto textOff = Point(ammoIconWidth - .5 * ammoIconHeight, .5 * (ammoIconHeight - font.Height()));
+	auto iconOff = Point(.5 * ammoIconHeight, .5 * ammoIconHeight);
+	const double iconCenterX = (ammoBox.Right() + ammoBox.Left()) / 2.;
+	for(const auto &it : ammo)
+	{
+		pos.Y() -= ammoIconHeight;
+		if(pos.Y() < ammoBox.Top() + ammoPad)
+			break;
```

Comment:
Could use a small comment
I'm not sure what the logic is here, if there are too many ammo types we dont show some?
Shouldnt we break when we reach the bottom and not the top? Maybe I'm not properly understanding what this does exactly

Reply:
Yes, don't try to draw more icons if we reach the top of the box allocated for ammo icons. Drawing further could potentially overlap with the flagship status display. I suppose this wouldn't actually be that bad, and would perhaps be more clear as to what's happening than not drawing additional icons.

The icons are drawn from the bottom up, if you only have a single secondary weapon, it's icon will be right at the bottom of the allowed area, not at the top.